src/NHibernate/Engine/TypedValue.cs

```diff
@@ -63,53 +80,66 @@ public class ParameterListComparer : IEqualityComparer<TypedValue>
 		{
 			public bool Equals(TypedValue x, TypedValue y)
 			{
-				if (y == null) return false;
+				if (ReferenceEquals(x, y))
+					return true;
```

Comment:
Shortcut added, although I do not expect it to be much useful, since its main usage is in query cache key comparison. Unless a not originally cached query instance is re-executed a second time (which is likely a code smell), typed values will be different instances.